Help Desk Sign In App

mjweiberg28 opened the pull request:
PR #1 for the Help Desk sign-in page

mjweiberg28 committed:
Establishing connection between Python and Sheets

Connecting a specified Google spreadsheet with the spreadsheet.py file. There is lots of commented out code for testing various ways to change data between the spreadsheet and Python.

mjweiberg28 committed:
Remove .idea from repo

mjweiberg28 committed:
Testing cases for flagged shifts

Adding cases that should trigger a 'flagged' status that the shift did not go as it should. If a case is flagged, it adds to 'flagged_items' sheet.

mjweiberg28 committed:
Making less Sheets API calls and increasing speed

Instead of making a read to the spreadsheet with each update to an individual cell, all values from certain sheets are read into a list of dictionaries. This lowers the chances of API exhaustion over the quota and increases the speed of the program. Creating the first pieces of the webpage(s).

mjweiberg28 committed:
overhaul 1.0

Molding project to be formatted similarly, but less complex, to the writing center project. More changes to come or will revert back to previous commit if this fails.

mjweiberg28 committed:
updating views

Updating the views files and their corresponding links. Successful basic navbar in local host.

mjweiberg28 committed:
HTML improvements with bootstrap

Editing multiple HTML pages, testing working links, bootstrap buttons and cards, etc.

mjweiberg28 committed:
testing more visuals

Testing more with Bootstrap (collapse, buttons, alerts). Visuals are still a work in progress

mjweiberg28 committed:
Minor edits to organization

mjweiberg28 committed:
changing views

mjweiberg28 committed:
Connected controler with views

The buttons to generate flagged shifts and reset the RFID scanner data are working to moderate success now.

mjweiberg28 committed:
changing button layouts

Making slight edits to HTML/CSS for appearances and layouts of buttons, adding clock in and out pages.

mjweiberg28 committed:
Beginning to add user info

Adding user HTML pages, beginning to check for (or input) users and usernames

mjweiberg28 committed:
Updating more HTML

mjweiberg28 committed:
Update shifts controller

Updating "bad" shifts generator for case if student clocks in (and out) when they were not scheduled for a shift at that time.

mjweiberg28 committed:
Add scanning ability to students page

Beginning stages of scanner functionality to be able to sign in as a student on the 'Student Employees' page

mjweiberg28 committed:
Basic sign in out connection between Python and Sheets

Adding basic abilities to send a sign-in or sign-out time from the website to python to the Google sheet.

mjweiberg28 committed:
Changes to forms and timeclock info

Various changes to shift_controller, views, and HTML forms for how time is stored/registered and sent between files.

mjweiberg28 committed:
Working Add Users Function

Adding users is working successfully, working now on adding rows to table of shifts when a sign-in occurs.

mjweiberg28 committed:
Updates to sign in/out

Signing in and out is currently working with the keyboard input and successfully using a post request to update the table.

mjweiberg28 committed:
Attempting to clear visual shifts each day without clearing data

Making changes to write to two different sheets, one that tracks the current day's shifts and the other that stores all shift data until Justin runs the program. Will likely backtrack on this and attempt another route.

c-schwarze commented:
@josiahtillman and @j-jasik This isn't finished yet, but this is kind of a check in. I'm going to take a look on Monday. If you guys could glance at this when you have a chance, that would be great!

c-schwarze commented on README.MD:
```diff
@@ -1,2 +1,34 @@
-a branch can be created.
\ No newline at end of file
+This project is designed to help the ITS Help Desk manager
+keep track of his students' work hours at the Help Desk
```
`his or her`

c-schwarze replied:
this happens a few times throughout this file

c-schwarze commented on README.MD:
```diff
@@ -1,2 +1,34 @@
+information into the 'hd_export' sheet of the Google
+Spreadsheet linked in 'spreadsheet.py'. With each clock in
+and out from their shifts, students will scan an RFID
+scanner and select whether they are signing in or out on
```
selecting whether they sign in and out is outdated.

c-schwarze commented on README.MD:
```diff
@@ -1,2 +1,34 @@
+* Web-pages for student sign in/out and HD manager's
+flagged-shift generator
+* Tying the RFID scanner information to code and testing
+* Any other testing or issues that arise
```
update the TODO list?

c-schwarze commented on app/__init__.py:
```diff
@@ -0,0 +1,10 @@
+
+app = Flask(__name__)
+app.config.from_object('config')
+
```
we will want to import sentry in this file. Although you may or may not have heard of it yet, it is something to keep in mind before the end of this project. Sentry allows us to capture error messages automatically.

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+from operator import itemgetter as i
+
+# Local
+from app import app
```
These imports look great 👍 nice job.

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+# 'hd_export' stores where the manager posts his expected shift schedule, typically for a two-week period
+hd_export = client.open('help_desk_sign_in').worksheet('hd_export')
+# 'hd_users' stores where the site inputs student username and card ID info for shift comparisons
+hd_users = client.open('help_desk_sign_in').worksheet('hd_users')
```
20-30, you call `client.open('help_desk_sign_in')` multiple times. Maybe just set that to a variable once, then reference it each time you need it. Might even save API calls, but certainly is a micro speed efficiency.

c-schwarze replied:
20-30, it might also help to name these variables so you know they are gsheets? gsheet_flagged_items, gsheet_rfid_input, etc.

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+            time = time[1:]
+        # try:
+        scan_list = rfid_input.range(len(rfid_input.get_all_records()) + 2, 1, len(rfid_input.get_all_records()) + 2, 4)
+        # except APIError as api_error:
```
we talked about this try/except earlier. Might be worth putting a TODO, so you don't forget

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+    def student_time_clock(self, card_id):
+        timestamp = datetime.now()
+        time = timestamp.strftime('%I:%M %p')
+        if time[0] == '0':
```
this if seems not needed? My understanding is you are removing the leading 0, if its there. You can probably find a strftime that better suites you than %I.

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+                scan_list[0].value = user['Username']
+                scan_list[1].value = timestamp.strftime('%x')
+                scan_list[2].value = time
+                rfid_input.update_cells(scan_list)
```
61-70: This is a bit confusing due to not knowing what `scan_list[0].value` and the others are. Might be worth some comments to help this out.

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+        timestamp = datetime.now()
+        time = timestamp.strftime('%I:%M %p')
+        if time[0] == '0':
+            time = time[1:]
```
75/76: came comment as earlier

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+                day_list[0].value = user['Username']
+                day_list[1].value = timestamp.strftime('%x')
+                day_list[2].value = time
+                current_day.update_cells(day_list)
```
72-92: seems too similar to 48-70?

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+    # refreshes the list of dictionaries of the hd_users sheet, allowing the users page to be up to date
+    def users_list(self):
+        users_list = hd_users.get_all_records()
+        return users_list
```
combine lines 98/99. `return hd_users.get_all_records()` This can also be done on 103/104 and 108/109

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+    # adds user info to the 'hd_users' sheet
+    def add_users(self, student_name, username, card_id):
+        self.users = hd_users.get_all_records()
+        cell_list = hd_users.range(len(self.users)+2, 1, len(self.users)+2, 3)
```
114: very confusing

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+
+    # called by multi_key_sort to compare and sort multiple keys in a dictionary
+    def cmp(self, a, b):
+        return (a > b) - (a < b)
```
120-122, you can put this method inside of the other method. This can be helpful for these types of "helper" methods, that only get used once inside a very specific method.

c-schwarze replied:
you actually do this for the `comparer` method already (probably copy pasted)

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+        d = datetime.strptime(str1, '%H:%M')
+        if d.strftime('%I:%M %p')[0] == '0':
+            return d.strftime('%I:%M %p')[1:]
+        return d.strftime('%I:%M %p')
```
methods `convert12` and `convert24` could probably be the same method with a parameter? Something to consider at least.

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+        for cell in cell_reset:
+            cell.value = ''
+        flagged_items.update_cells(cell_reset)
+        return
```
having an empty return at the end of methods is a bad practice. Either don't return, or return something useful that might be used. This is in a few places

Java definitely does this, but python doesn't have to have returns

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+        for item in hd_input:
+            # empty shifts set as 'zz - empty' to put them at the bottom of the alphabetical sort
+            if item['Employee Name'] == '':
+                item['Employee Name'] = 'zz - empty'
```
seems weird to set them as `zz - empty`. I like the idea though. Maybe use symbols instead?

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+            # empty shifts set as 'zz - empty' to put them at the bottom of the alphabetical sort
+            if item['Employee Name'] == '':
+                item['Employee Name'] = 'zz - empty'
+            if item['Start Time'][-1] != 'M':
```
201 and 203, you are looking for the M in AM/PM, right? Seems like a goofy way to check for this.

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+        for item in scan_input:
+            # TODO: Comparing user-names in scan_input to names in hd_input
+            # TODO: Inefficient and non-maintainable code currently, will need to fix later
+            # TODO: Current setup exists for comparison in special cases
```
Nice job with the TODOs. I would highly recommend doing more of this for all projects. 👍

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+        # determines what row of 'flagged_items' to update for a new flagged item, beginning at row 2
+        flag_count = 2
+        # tracks the correct row number for RFID time in and out, taking into account possible extra shifts
+        p = 0
```
`p` not the best variable name. Maybe be more descriptive?

c-schwarze commented on app/shifts_controller.py:
```diff
@@ -0,0 +1,363 @@
+                    time_in = datetime.strptime(scan_input[p]['Date'] + scan_input[p]['In'], '%x%H:%M')
+
+            # case: student forgets to clock out (time out value is empty)
+            # forgetting to clock in but then clocking out will be read by the scanner as forgetting to clock out
```
293, thats a great comment and thing to note

c-schwarze commented on app/static/hd_shifts.css:
```diff
@@ -0,0 +1,119 @@
+    margin: 0 0 12px;
+}
+
+.wc-margins {
```
wc?

c-schwarze commented on app/static/shift_buttons.js:
```diff
@@ -0,0 +1,3 @@
+// nothing here currently
+// all JS code is currently in the HTML files for student_index, staff_index, and users_index
+// could be moved here instead of how it currently resides in HTML files as <script>
```
definitely worth doing this. Moving the JS out of HTML is a good practice.

c-schwarze commented on app/templates/help_desk_base.html:
```diff
@@ -0,0 +1,46 @@
+
+{% block html_attribs %}
+    <html lang="en">
+{% endblock %}
```
You shouldn't need to set blocks, if you aren't intending on replacing them elsewhere. This might apply to other spots

c-schwarze reviewed (changes requested):
Overall this is great!

most things i commented on are micro changes or formatting.

There are a few things that I didn't go over too deeply, but we can do more code reviews down the road

j-jasik commented on app/templates/users_index.html:
```diff
@@ -0,0 +1,66 @@
+                    </tr>
+                </thead>
+                <tbody id="user-tbody">
+                    {% include 'users_table.html' %}
```
Is there a reason we include the table this way? Could you just build it here and get rid of the other file?

mjweiberg28 replied:
I guess I have it included like this so that when a student scans their ID and clocks in/out, only the table portion of the page is reloaded and not the entire page. The 'verify_scanner' method from my 'views.py' renders only 'users_table.html'. Not sure how much of a difference it makes though, or if there is a better way to be only reloading the table and not the whole page.
Edit: I thought this comment was on the similar way that the student_table.html is set up. The above applies to that. 'users_index.html' and 'users_table.html' no longer exist because Caleb recommended a different route for adding users.

j-jasik reviewed (commented):
I agree it's looking good! I just had one small comment/question to think about.

mjweiberg28 committed:
Updating after PR 1

Lots of minor updates to code in response to the comments from pull request 1

mjweiberg28 committed:
Minor edits to limit API reads

Editing how a few of the sheets are being called in shifts_controller.py in order to prevent the "resource exhausted" error with the Sheets API.

mjweiberg28 committed:
Table display

Improved efficiency of displaying the current day's shifts on the main student time-clock page. Also beginning to implement try/catch.

mjweiberg28 committed:
Try/except improvements

Adding fade actions to a few Bootstrap alerts and getting a few of the try/except's working properly. Still in progress

mjweiberg28 committed:
Try/except Bootstrap Alerts

Alerts for successful/unsuccessful timeclock entries and shift processor appear to be fully working. Alert for exhausted server also works when Sheets API is over-worked.

mjweiberg28 committed:
Creating a Help Page

Adding a FAQ-style help page for the HD manager to view instructions on how the site works or answer basic questions about things on the site.

mjweiberg28 committed:
Eliminating ['Username']

Getting rid of user['Username'] and shift['Username'] in hd_users and scan_input sheets, respectively. It was unnecessary and causing extra API calls.
